src/org/traccar/notification/NotificationFormatter.java

```diff
@@ -115,4 +118,20 @@ private static String formatMessage(VelocityContext vc, Long userId, Event event
         return writer.toString();
     }
 
+    public static Map<String, String> buildData(long userId, Event event, Position position) {
+        Map<String, String> data = new HashMap<>();
+        try {
+            if (event != null) {
+                String eventJson = Context.getObjectMapper().writeValueAsString(event);
+                data.put("eventJson", eventJson);
+            }
+            if (position != null) {
+                String positionJson = Context.getObjectMapper().writeValueAsString(position);
+                data.put("positionJson", positionJson);
+            }
+        } catch (JsonProcessingException e) {
```

Comment:
I would throw this exception instead.

Reply:
copied from AsyncSocket::sendData. It might be better to send a dataless notification, than nothing IMHO. Maybe I can refactor NotificatorFirebase so that if it gets an empty data map (or null?) then it falls back to the notification without data?